bpo-40915: Fix mmap resize bugs on Windows (GH-29213)

tjguk opened the pull request:
[bpo-40915](https://bugs.python.org/issue40915): Fix mmap resize bugs on Windows

The current implementation of `mmap.resize` on Windows contains several subtle bugs. This PR, principally using code contributed by eryksun, addresses those bugs and adds corresponding tests.

I've added some doc changes, but I'm not convinced that they should go in, since the cases addressed are relatively niche.

<!-- issue-number: [bpo-40915](https://bugs.python.org/issue40915) -->
https://bugs.python.org/issue40915
<!-- /issue-number -->

tjguk committed:
On Windows, if a named mmap section is held by more than one handle, no handle can be resized

tjguk committed:
A pagefile-backed mmap can now be resized (under the cover it's being recreated along with its data)

tjguk committed:
Merge branch 'main' into issue40915-mmap-resize

tjguk committed:
Make sure that resizing includes the offset when calculating the new file extent

tjguk committed:
Add & tweak resize tests for Windows:

* Check that a resize against the pagefile succeeds and copies as much data as will fit
* Check that a resize against a named file correctly fails where another mapping is still open

tjguk committed:
Add useful commentary to the new resize logic

tjguk committed:
Add some tentative docs to reflect the new changes

zooba commented on Modules/mmapmodule.c:
```diff
@@ -374,16 +375,36 @@ is_writable(mmap_object *self)
+            }
+        }
+    }
+#endif (MS_WINDOWS)
```
```suggestion
#endif /* MS_WINDOWS */
```

zooba commented on Modules/mmapmodule.c:
```diff
@@ -374,16 +375,36 @@ is_writable(mmap_object *self)
 is_resizeable(mmap_object *self)
 {
+#ifdef MS_WINDOWS
+    /* a named file mapping that's open more than once can't be resized */
```
Won't we find this out later on when we try to resize it? I'd rather not be pulling in internal functions at all if we can avoid it.

zooba replied:
Or alternatively, can we try to resize it to its current size?

tjguk replied:
Reworked to remove the up-front check. Instead we correctly handle the situation below by leaving the mapping intact but allowing an exception to bubble up. Added a specific test for this situation.

Not sure whether a different exception might be better, but then what?

tjguk committed:
Add a specific test to check that resizing a named mapping fails with an exception but retains the mapping
Remove the specific pre-check on resizing a named section but specifically account for it later when resizing

zooba commented on Modules/mmapmodule.c:
```diff
@@ -503,51 +505,96 @@ mmap_resize_method(mmap_object *self,
+        that creating a new named file mapping failed and the mapping handle
+        points to an existing mapping of the same name
+        */
+        if ((GetLastError() != ERROR_ALREADY_EXISTS) &&
```
Should this compare against `error` rather than calling `GetLastError()` again?

In fact, it definitely should, because `GetLastError()` _is not_ guaranteed to return 0 when the last call was successful - it only updates on errors.

```suggestion
        if (error && error != ERROR_ALREADY_EXISTS) &&
```

(I'm not 100% sure whether we want to unmap in the success case or just any case besides ERROR_ALREADY_EXISTS. Trusting that you know.)

eryksun replied:
> In fact, it definitely should, because `GetLastError()` _is not_ guaranteed to return 0 when the last call was successful - it only updates on errors.

For create calls that use open-always semantics, such as `CreateFile()` and `CreateFileMapping()`, on success the last error is set to either 0 or `ERROR_ALREADY_EXISTS`. In our case, we need to handle `ERROR_ALREADY_EXISTS` as an error, and close the opened handle.

tjguk replied:
I was trying to be clever and avoid referring to `error` when it might not have been defined. But of course it's always defined (at least to zero if nothing's got in the way) so all you need is an `if (error != ERROR_ALREADY_EXISTS)`. Which I've now done.

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -503,51 +505,96 @@ mmap_resize_method(mmap_object *self,
-        return NULL;
+        else {
+            error = GetLastError();
+        }
```
For the `ERROR_ALREADY_EXISTS` case, for which `self->map_handle` will be a valid handle, this is missing `CloseHandle(self->map_handle)` and `self->map_handle = NULL`.

Resizing shouldn't be allowed if a named section has multiple handles, which we can try to check directly or indirectly (see below). It shouldn't render the mmap useless when it fails, i.e. the existing `self->data` view shouldn't be unmapped. We can't reliably recover after unmapping the view and re-opening the section with `ERROR_ALREADY_EXISTS`. If there really was only one handle, it most likely means an unrelated section object was created in between calling `CloseHandle()` and `CreateFileMapping()`.

It's probably good enough to call `OpenFileMapping()` immediately after `CloseHandle()`. If it succeeds, then fail the resize. The window of time between the `CloseHandle()` and `OpenFileMapping()` calls is short enough that we can reasonably assume the opened handle is for our section object. It's still not perfect, though.

tjguk replied:
Thanks @eryksun (and thanks obviously for the original code on which this patch is largely based). Thinking aloud here for a moment::

If there is genuinely more than one handle against the same named section to start with, the `CreateFileMapping` will both set `ERROR_ALREADY EXISTS` as well as returning a *valid* handle to the named section. We raise an exception but leave a mapping in place which points to the same data and has the original size. (This is what the current code does). We never unmap the view and the new handle is good, so no need to close it and invalidate the mmap object.

If there is only one handle to a named section, we might get a race condition if someone someone creates an identically-named but unconnected handle in between the `Close` and the `Create`. I think that's what your point is?

I did hope that the handle returned by a second `CreateFileMapping` against the same named section would return the same `HANDLE` and so give us a means of detecting whether we really had the same named section. But, experimenting, it doesn't. So that brings us back to the `OpenFileMapping` you suggest as quickly as possible. I'll give that a try.

eryksun replied:
Here's an untested rewrite:

```C
        /* close the file mapping */
        CloseHandle(self->map_handle);
        /* if the file mapping still exists, it cannot be resized. */
        if (self->tagname) {
            self->map_handle = OpenFileMapping(FILE_MAP_WRITE, FALSE,
                                    self->tagname);
            if (self->map_handle) {
                PyErr_SetFromWindowsErr(ERROR_USER_MAPPED_FILE);
                return NULL;
            }
        } else {
            self->map_handle = NULL;
        }

        /* if it's not the paging file, unmap the view and resize the file */
        if (self->file_handle != INVALID_HANDLE_VALUE) {
            if (!UnmapViewOfFile(self->data)) {
                return PyErr_SetFromWindowsErr(GetLastError());
            };
            self->data = NULL;
            /* resize the file */
            if (!SetFilePointerEx(self->file_handle, max_size, NULL,
                FILE_BEGIN) ||
                !SetEndOfFile(self->file_handle)) {
                /* resizing failed. try to remap the file */
                error = GetLastError();
                new_size = max_size.QuadPart = self->size;
            }
        }
        
        /* create a new file mapping and map a new view */
        /* FIXME: call CreateFileMappingW with wchar_t tagname */
        self->map_handle = CreateFileMapping(
            self->file_handle,
            NULL,
            PAGE_READWRITE,
            max_size.HighPart,
            max_size.LowPart,
            self->tagname);

        error = GetLastError();
        if (error == ERROR_ALREADY_EXISTS) {
            CloseHandle(self->map_handle);
            self->map_handle = NULL;
        }
        else if (self->map_handle != NULL) {
            self->data = MapViewOfFile(self->map_handle,
                FILE_MAP_WRITE,
                offset.HighPart,
                offset.LowPart,
                new_size);
            if (self->data != NULL) {
                /* copy the old view if using the paging file */
                if (self->file_handle == INVALID_HANDLE_VALUE) {
                    memcpy(self->data, old_data,
                           self->size < new_size ? self->size : new_size);
                    if (!UnmapViewOfFile(old_data)) {
                        error = GetLastError();
                    }
                }
                self->size = new_size;
            }
            else {
                error = GetLastError();
                CloseHandle(self->map_handle);
                self->map_handle = NULL;
            }
        }
        
        if (error) {
            PyErr_SetFromWindowsErr(error);
            return NULL;
        }
        Py_RETURN_NONE;
```

tjguk replied:
Thanks, @eryksun. Thing is: that changes the logic when two handles are open against the same underlying file. In the previous code, that would raise 1224 - ERROR_USER_MAPPED_FILE. But, in this version, the `error = GetLastError();` under the comment "Resizing failed..." is then overtaken by the unconditional `error = GetLastError();` after the `CreateFileMapping` call below. So the resize fails, but no exception is raised so the user never knows.

Hopefully I can some with something variant which covers all the bases

eryksun replied:
I noticed that. The problem that needs to be addressed is raising a chained exception based on two possible errors, `file_resize_error` and `map_error`. If resizing the file fails, we need to try to remap the existing file. But remapping may also fail. This should raise a chained exception that includes both errors.

eryksun replied:
I think the final error handling would go like this -- untested:

```C
    if (map_error || file_resize_error) {
        if (file_resize_error) {
            PyErr_SetFromWindowsErr(file_resize_error);
        }
        if (map_error) {
            PyObject *exc, *val, *tb;
            PyErr_Fetch(&exc, &val, &tb);
            PyErr_SetFromWindowsErr(map_error);
            if (exc != NULL) {
                _PyErr_ChainExceptions(exc, val, tb);
            }
        }
        return NULL;
    }
```

tjguk replied:
What I've implemented is a compromise: rather than entering the world of chained exceptions, I've simply kept track of the file resize and raised it as an exception even if no other error is present. For my part I think that's good enough for now.

zooba commented on Modules/mmapmodule.c:
```diff
@@ -503,51 +505,96 @@ mmap_resize_method(mmap_object *self,
+        */
+        if ((GetLastError() != ERROR_ALREADY_EXISTS) &&
+            (self->file_handle == INVALID_HANDLE_VALUE)) {
+            if(!UnmapViewOfFile(old_data)) {
```
If this succeeds, do we want to return `None` (in which case `error` needs to be cleared before line 593)? Or still raise the previous error (that is, this is part of a "finally" block)?

tjguk replied:
The intention is that the error (essentially `FileExistsError`) be raised as an indication to the user that the resize didn't work.

zooba reviewed (approved):
Nice change. I still have a couple more questions, but if you're happy with your answers, don't wait on me before merging.

akulakov commented on Doc/library/mmap.rst:
```diff
@@ -256,6 +256,14 @@ To map anonymous memory, -1 should be passed as the fileno along with the length
       with :const:`ACCESS_READ` or :const:`ACCESS_COPY`, resizing the map will
       raise a :exc:`TypeError` exception.
 
+      **On Windows**: Resizing the map will fail if there are other maps against
```
It may be good say that OSError is raised to be consistent with paragraph above that states which error is raised.

tjguk replied:
Thanks. Done

tjguk committed:
Use the existing `error` value as `GetLastError` might return a different value by this point

tjguk committed:
Be specific about the exception type raised

tjguk committed:
Pick up an error when resizing the file even if everything else succeeds

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -503,51 +505,110 @@ mmap_resize_method(mmap_object *self,
     }
 
     {
+        /*
+        To resize an mmap on Windows:
+
+        - Close the existing mapping
+        - If the mapping is backed to a named file:
+            unmap the view, clear the data, and resize the file
+            If the file can't be resized (eg because it has other mapped references
+            to it) then let the mapping be recreated at the original size and set
+            an error code so an exception will be raised.
+        - Create a new mapping of the relevant size to the same file
+        - Map a new view of the resized file
+        - If the mapping is backed by the pagefile:
+            copy any previous data into the new mapped area
+            unmap the original view which will release the memory
+        */
```
I don't think this extended comment is needed. It's a burden to keep it in sync if the implementation is changed later. Maybe clarify comments at the various steps if the operation or intention is unclear.

tjguk replied:
Fair enough. I think it reflects a little the thought process I was going through when trying to juggle all the different combinations. I'll remove it from here and consider adding other comments elsewhere

eryksun commented on Modules/mmapmodule.c:
```diff
@@ -503,51 +505,110 @@ mmap_resize_method(mmap_object *self,
-                                                new_size);
+
+        error = GetLastError();
+        if (error == ERROR_ALREADY_EXISTS) {
+            CloseHandle(self->map_handle);
+            self->map_handle = NULL;
```
This warrants a comment that `ERROR_ALREADY_EXISTS` indicates that someone else created a mapping with the given name in the time that it took to unmap the previous view and resize the file.

tjguk replied:
Done

eryksun commented on Lib/test/test_mmap.py:
```diff
@@ -796,6 +796,80 @@ def test_madvise(self):
+        """
+        start_size = 2 * PAGESIZE
+        reduced_size = PAGESIZE
+        tagname = "TEST"
```
I'm sorry to have missed this before you merged. Naming kernel objects with generic names like "foo" and "TEST" is a bad idea because the namespace is shared by every process in the current session. It's a serious problem in all of the tests that use `tagname`. If the name is an existing Section object (i.e. file mapping), `CreateFileMapping()` will succeed with the last error set to `ERROR_ALREADY_EXISTS`. In many cases, the Section object likely refers to an unrelated file, but we can't avoid this, or even know about the problem, because `mmap` doesn't currently support exclusive ("x") creation. This and a few other cases really need to be addressed in the constructor, but that's a separate issue.

tjguk replied:
Thanks @eryksun  -- good catch. Funnily enough, I'm normally a little over-obsessive with test names, generating them randomly to descrease any chance of a "lucky name". This time I've gone the other way! I'll run up a quick patch to fix this.
